Polish text editor mode

jasmussen opened the pull request:
This PR adds some polish to the text editor mode:

It adds a dummy formatting toolbar. Not yet sticky. 

It makes the editor textarea 100% full width and uses paddings to center it. I feel this is somewhat controversial, so please share your concerns here. My reason for doing this is that it allows you to click _anywhere_ on the page to set focus on the text editor. Since we are using a very boundless and full-bleed look for this, I feel that is important for accessibility.

I was also fiddling a bit with CSS classes and tags to use in the dummy markup, please share your thoughts here as well, thanks. 

Finally, Andrew brought up media queries in another PR — since the padding trick (if you should all find it kosher) relies on a few of these, might as well bring that up here. Should we bring in a mixin to handle this? Should we use a whitelist of breakpoints, or just assign them as valuable and use the existing WordPress ones as the baseline?

Thanks for your time!

jasmussen committed:
Polish the editor bar buttons.

Just a tiny 1px fix.

jasmussen committed:
Add dummy HTML formatting bar.

jasmussen committed:
Use padding instead of margins for texteditor.

This allows you to click anywhere in the white area to set focus on the textield. Though this is a kludgier centering experience, it seems to work pretty well with the breakpoints I added. Thoughts welcome.

jasmussen commented:
Question: right now the content in text editor mode is one big blob. Will that change when we bring in autop, or are there different plans here?

youknowriad commented:
> Question: right now the content in text editor mode is one big blob. Will that change when we bring in autop, or are there different plans here?

Good question! Right now if we add spacing between blocks, the parse will consider those as "unparsed" html and thus creates a "freeform" block containing these spaces.

And if we add spaces after the opening comment or before the closing comment, the parser will consider these spaces as part of the block's content.

Maybe we should trip the content between and inside blocks which would allow us to serialize the blocks in a more readable format.

jasmussen commented:
I don't have too strong opinions on what happens behind the scenes, so to speak —  my trust in you is absolute. But it would be nice if you could edit a post in the text editor without having to treat it as one big blob, even if we do that through "magic tricks". 

One of the common use cases for the text editor, is to fix tag soup from the visual editor, and though we mean to create less tag soup with the new editor you never know what might be pasted in here and there. So anything we could do to allow roundtrips to be easier on the eyes would be really nice.

aduth commented:
>Maybe we should trim the content between and inside blocks which would allow us to serialize the blocks in a more readable format.

This might be alright, depending on whether we'd need to support blocks that must have leading or tailing whitespace (probably not?).

Also probably fine to ignore blocks parsed as freeform with only whitespace?

aduth commented on editor/modes/text-editor/index.js:
```diff
@@ -11,12 +11,32 @@ import './style.scss';
-		/>
+		<div>
+			<header className="editor-text-editor__formatting">
+				<div>
```
Can we lose the `div` and apply the padding to the `.editor-text-editor__formatting` directly? In any case we should try to avoid applying styles via element selectors, considering there's a non-zero likelihood one of these buttons may eventually contain their own `div` descendants.

aduth commented on editor/modes/text-editor/style.scss:
```diff
@@ -1,12 +1,98 @@
+		outline: none;
+	}
+
+	.editor-text-editor__bold {
```
Do these selectors need to be nested? One of the advantages of BEM is to not need excessive nesting since the names are self-describing and unique. Usually becomes an issue only when fighting for specificity.

aduth reviewed (commented):
I'm fine with the idea of using padding to allow click-anywhere focus.

joyously commented:
> It makes the editor textarea 100% full width and uses paddings to center it. 

As long as the parent container is getting its width from the theme's editor-style.css (or $content-width?).

jasmussen commented:
> As long as the parent container is getting its width from the theme's editor-style.css (or $content-width?).

This would be for the text/HTML mode, where the editor style isn't loaded.

jasmussen committed:
Lose the div.

jasmussen committed:
Remove a little specificity.

jasmussen commented:
I addressed the feedback in this thread. Two remaining tasks are:

- use a mixin + variables for breakpoints
- make the text editor formatting bar sticky

I would like to address both of these in a separate PR to keep things flowing smoothly, so if I can get a thumbs up and merge this then I'll proceed, thanks!

youknowriad reviewed (approved):
Tested and LGTM 👍

jasmussen commented:
Thank you!
